File: src/test/java/org/example/eventhub/service/UserServiceTest.java
package org.example.eventhub.service;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

import org.example.eventhub.dto.security.RegisterRequest;
import org.example.eventhub.dto.user.*;
import org.example.eventhub.entity.User;
import org.example.eventhub.enums.Role;
import org.example.eventhub.exception.UserAlreadyExistsException;
import org.example.eventhub.exception.UserNotFoundException;
import org.example.eventhub.mapper.UserMapper;
import org.example.eventhub.repository.UserRepository;
import org.example.eventhub.specification.UserSpecification;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.data.domain.*;
import org.springframework.data.jpa.domain.Specification;

class UserServiceTest {

    private static final Long EXISTING_ID = 1L;
    private static final Long NON_EXISTING_ID = 999L;
    private static final String EXISTING_USERNAME = "john_doe";
    private static final String EXISTING_EMAIL = "john@example.com";
    private static final String NEW_USERNAME = "jane_doe";
    private static final String NEW_EMAIL = "jane@example.com";

    private final UserRepository repository = mock(UserRepository.class);
    private final UserMapper mapper = mock(UserMapper.class);
    private final UserSpecification realSpecification = new UserSpecification();

    private final UserService userService = new UserService(repository, mapper, realSpecification);

    private User user;
    private RegisterRequest createRequest;
    private UserUpdateRequest updateRequest;
    private UserResponseLong responseLong;
    private UserResponseShort responseShort;
    private Pageable pageable;

    @BeforeEach
    void setUp() {
        user = new User();
        user.setId(EXISTING_ID);
        user.setUsername(EXISTING_USERNAME);
        user.setEmail(EXISTING_EMAIL);
        user.setPassword("oldPassword");

        createRequest = new RegisterRequest("new_user", "new@example.com", "password123");

        updateRequest = new UserUpdateRequest(NEW_USERNAME, NEW_EMAIL, "newPassword");

        responseLong = new UserResponseLong(
                EXISTING_ID, EXISTING_USERNAME, EXISTING_EMAIL, Role.USER, LocalDateTime.now(), 10);

        responseShort = new UserResponseShort(EXISTING_USERNAME);

        pageable = PageRequest.of(0, 10);
    }

    // ====================== createUser ======================

    @Test
    @DisplayName("createUser: успешно создаёт нового пользователя")
    void createUser_success() {
        when(repository.findByUsername(createRequest.username)).thenReturn(Optional.empty());
        when(repository.findByEmail(createRequest.email)).thenReturn(Optional.empty());
        when(mapper.toEntity(createRequest)).thenReturn(user);
        when(repository.save(user)).thenReturn(user);
        when(mapper.toLongDto(user)).thenReturn(responseLong);

        UserResponseLong result = userService.createUser(createRequest);

        assertEquals(responseLong, result);
        verify(repository).save(user);
    }

    @Test
    @DisplayName("createUser: бросает исключение, если username уже существует")
    void createUser_usernameAlreadyExists() {
        when(repository.findByUsername(createRequest.username)).thenReturn(Optional.of(user));

        UserAlreadyExistsException ex =
                assertThrows(UserAlreadyExistsException.class, () -> userService.createUser(createRequest));

        assertEquals("Пользователь с username " + createRequest.username + " уже существует", ex.getMessage());
        verify(repository, never()).save(any());
    }

    @Test
    @DisplayName("createUser: бросает исключение, если email уже существует")
    void createUser_emailAlreadyExists() {
        when(repository.findByUsername(createRequest.username)).thenReturn(Optional.empty());
        when(repository.findByEmail(createRequest.email)).thenReturn(Optional.of(user));

        UserAlreadyExistsException ex =
                assertThrows(UserAlreadyExistsException.class, () -> userService.createUser(createRequest));

        assertEquals("Пользователь с email " + createRequest.email + " уже существует", ex.getMessage());
        verify(repository, never()).save(any());
    }

    // ====================== getAllUsers ======================

    @Test
    @DisplayName("getAllUsers: возвращает страницу Short DTO с фильтром")
    void getAllUsers_returnsPage() {
        Page<User> page = new PageImpl<>(List.of(user), pageable, 1);
        UserFilter filter = new UserFilter("john", null, null, null, null, null);

        when(repository.findAll(any(Specification.class), eq(pageable))).thenReturn(page);
        when(mapper.toShortDto(user)).thenReturn(responseShort);

        Page<UserResponseShort> result = userService.getAllUsers(pageable, filter);

        assertEquals(1, result.getTotalElements());
        assertEquals(responseShort, result.getContent().get(0));
        verify(repository).findAll(any(Specification.class), eq(pageable));
    }

    @Test
    @DisplayName("getAllUsers: возвращает пустую страницу")
    void getAllUsers_emptyPage() {
        Page<User> emptyPage = Page.empty(pageable);

        when(repository.findAll(any(Specification.class), eq(pageable))).thenReturn(emptyPage);

        Page<UserResponseShort> result = userService.getAllUsers(pageable, null);

        assertTrue(result.isEmpty());
    }

    // ====================== getUserById ======================

    @Test
    @DisplayName("getUserById: возвращает Long DTO при существующем пользователе")
    void getUserById_success() {
        when(repository.findById(EXISTING_ID)).thenReturn(Optional.of(user));
        when(mapper.toLongDto(user)).thenReturn(responseLong);

        UserResponseLong result = userService.getUserById(EXISTING_ID);

        assertEquals(responseLong, result);
    }

    @Test
    @DisplayName("getUserById: бросает исключение, если пользователь не найден")
    void getUserById_notFound() {
        when(repository.findById(NON_EXISTING_ID)).thenReturn(Optional.empty());

        UserNotFoundException ex =
                assertThrows(UserNotFoundException.class, () -> userService.getUserById(NON_EXISTING_ID));

        assertEquals("Пользователь с id " + NON_EXISTING_ID + " не найден", ex.getMessage());
    }

    // ====================== getUserByUsername ======================

    @Test
    @DisplayName("getUserByUsername: возвращает Long DTO по username")
    void getUserByUsername_success() {
        when(repository.findByUsername(EXISTING_USERNAME)).thenReturn(Optional.of(user));
        when(mapper.toLongDto(user)).thenReturn(responseLong);

        UserResponseLong result = userService.getUserByUsername(EXISTING_USERNAME);

        assertEquals(responseLong, result);
    }

    @Test
    @DisplayName("getUserByUsername: бросает исключение, если пользователь не найден")
    void getUserByUsername_notFound() {
        when(repository.findByUsername("unknown")).thenReturn(Optional.empty());

        UserNotFoundException ex =
                assertThrows(UserNotFoundException.class, () -> userService.getUserByUsername("unknown"));

        assertEquals("Пользователь с username unknown не найден", ex.getMessage());
    }

    // ====================== updateUser ======================

    @Test
    @DisplayName("updateUser: успешно обновляет все поля")
    void updateUser_success_allFields() {
        when(repository.findById(EXISTING_ID)).thenReturn(Optional.of(user));
        when(repository.findByUsername(NEW_USERNAME)).thenReturn(Optional.empty());
        when(repository.findByEmail(NEW_EMAIL)).thenReturn(Optional.empty());
        when(repository.save(user)).thenReturn(user);
        when(mapper.toLongDto(user)).thenReturn(responseLong);

        UserResponseLong result = userService.updateUser(EXISTING_ID, updateRequest);

        assertEquals(NEW_USERNAME, user.getUsername());
        assertEquals(NEW_EMAIL, user.getEmail());
        assertEquals("newPassword", user.getPassword());
        assertEquals(responseLong, result);
        verify(repository).save(user);
    }

    @Test
    @DisplayName("updateUser: обновляет только email, если username не меняется")
    void updateUser_onlyEmailChanged() {
        UserUpdateRequest partial = new UserUpdateRequest(null, NEW_EMAIL, null);

        when(repository.findById(EXISTING_ID)).thenReturn(Optional.of(user));
        when(repository.findByEmail(NEW_EMAIL)).thenReturn(Optional.empty());
        when(repository.save(user)).thenReturn(user);
        when(mapper.toLongDto(user)).thenReturn(responseLong);

        userService.updateUser(EXISTING_ID, partial);

        assertEquals(EXISTING_USERNAME, user.getUsername());
        assertEquals(NEW_EMAIL, user.getEmail());
    }

    @Test
    @DisplayName("updateUser: бросает исключение при попытке взять занятый username")
    void updateUser_usernameAlreadyTaken() {
        UserUpdateRequest badRequest = new UserUpdateRequest("taken_user", null, null);
        User anotherUser = new User();
        anotherUser.setUsername("taken_user");

        when(repository.findById(EXISTING_ID)).thenReturn(Optional.of(user));
        when(repository.findByUsername("taken_user")).thenReturn(Optional.of(anotherUser));

        assertThrows(UserAlreadyExistsException.class, () -> userService.updateUser(EXISTING_ID, badRequest));
    }

    // ====================== deleteUser ======================

    @Test
    @DisplayName("deleteUser: успешно удаляет существующего пользователя")
    void deleteUser_success() {
        when(repository.findById(EXISTING_ID)).thenReturn(Optional.of(user));

        userService.deleteUser(EXISTING_ID);

        verify(repository).deleteById(EXISTING_ID);
    }

    @Test
    @DisplayName("deleteUser: бросает исключение, если пользователь не найден")
    void deleteUser_notFound() {
        when(repository.findById(NON_EXISTING_ID)).thenReturn(Optional.empty());

        assertThrows(UserNotFoundException.class, () -> userService.deleteUser(NON_EXISTING_ID));
    }
}

File: src/test/kotlin/org/example/eventhub/service/UserServiceTest.kt
package org.example.eventhub.service

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.shouldBe
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.example.eventhub.dto.security.RegisterRequest
import org.example.eventhub.dto.user.UserFilter
import org.example.eventhub.dto.user.UserResponseLong
import org.example.eventhub.dto.user.UserResponseShort
import org.example.eventhub.dto.user.UserUpdateRequest
import org.example.eventhub.mapper.UserMapper
import org.example.eventhub.repository.UserRepository
import org.example.eventhub.specification.UserSpecification
import org.example.eventhub.entity.User
import org.example.eventhub.enums.Role
import org.example.eventhub.exception.UserAlreadyExistsException
import org.example.eventhub.exception.UserNotFoundException
import org.springframework.data.domain.PageImpl
import org.springframework.data.domain.PageRequest
import org.springframework.data.jpa.domain.Specification
import org.springframework.security.crypto.password.PasswordEncoder
import java.time.LocalDateTime
import java.util.Optional

class UserServiceTest : StringSpec({

    val repository = mockk<UserRepository>()
    val mapper = mockk<UserMapper>()
    val specification = mockk<UserSpecification>()
    val passwordEncoder = mockk<PasswordEncoder>()

    val userService = UserService(repository, mapper, specification, passwordEncoder)

    val user = User(
        id = 1,
        username = "username",
        email = "email@email.com",
        password = "password",
        role = Role.USER
    )
    val userResponseLong = UserResponseLong(
        id = 1,
        username = "username",
        email = "email@email.com",
        role = Role.USER,
        createdAt = LocalDateTime.of(2021, 1, 1, 1, 1, 0),
        countOrganizedEvents = 0
    )
    val userResponseShort = UserResponseShort(
        username = "username",
    )
    val filter = UserFilter(null, null, null, null, null, null)
    val page = PageRequest.of(0, 10)


    every { mapper.toLongDto(user) } returns userResponseLong
    every { mapper.toShortDto(user) } returns userResponseShort

    every { passwordEncoder.encode(any()) } returns "password"

    every { repository.findByUsername("notExistingUsername") } returns Optional.empty()
    every { repository.findByEmail("notExisting@email.com") } returns Optional.empty()
    every { repository.findById(-1) } returns Optional.empty()

    every { repository.findByUsername("existingUsername") } returns Optional<User>.of(user)
    every { repository.findByEmail("existing@email.com") } returns Optional<User>.of(user)
    every { repository.findById(1) } returns Optional<User>.of(user)

    every { repository.save(any<User>()) } returns user


    //----------------createUser-----------------

    "createUser сохраняет пользователя в БД и возвращает UserResponseLong" {
        val dto = RegisterRequest(
            username = "notExistingUsername",
            email = "notExisting@email.com",
            password = "password"
        )
        every { mapper.toEntity(dto, "password") } returns user

        val response = userService.createUser(dto)

        response shouldBe userResponseLong

        verify(exactly = 1) { passwordEncoder.encode(user.password) }
        verify(exactly = 1) { repository.findByUsername("notExistingUsername") }
        verify(exactly = 1) { repository.findByEmail("notExisting@email.com") }
        verify(exactly = 1) { repository.save(user) }
    }

    "createUser выбрасывает исключение, если пользователь с таким же username существует" {
        val dto = RegisterRequest(
            username = "existingUsername",
            email = "notExisting@email.com",
            password = "password"
        )

        val ex = shouldThrow<UserAlreadyExistsException> {
            userService.createUser(dto)
        }

        ex.message shouldBe "Пользователь с username existingUsername уже существует"
    }

    "createUser выбрасывает исключение, если пользователь с таким же email существует" {
        val dto = RegisterRequest(
            username = "notExistingUsername",
            email = "existing@email.com",
            password = "password"
        )

        val ex = shouldThrow<UserAlreadyExistsException> {
            userService.createUser(dto)
        }

        ex.message shouldBe "Пользователь с email existing@email.com уже существует"
    }

    //----------------getAllUsers-----------------

    "getAllUsers возвращает список всех пользователей, если нет фильтра" {
        val userSpec = mockk<Specification<User>>()

        every { specification.withFilter(filter) } returns userSpec
        every { repository.findAll(userSpec, page) } returns PageImpl(listOf(user))

        val result = userService.getAllUsers(page, filter)

        result.content shouldBe listOf(userResponseShort)
    }

    //----------------getUserById-----------------

    "getUserById возвращает UserResponseLong, если пользователь есть" {
        val result = userService.getUserById(1)
        result shouldBe userResponseLong
    }

    "getUserById выбрасывает ошибку, если пользователя не существует" {
        val ex = shouldThrow<UserNotFoundException> {
            userService.getUserById(-1)
        }

        ex.message shouldBe "Пользователь с id -1 не найден"
    }

    //----------------getUserByUsername-----------------

    "getUserByUsername возвращает UserResponseLong, если пользователь есть" {
        val result = userService.getUserByUsername("existingUsername")
        result shouldBe userResponseLong
    }

    "getUserByUsername выбрасывает ошибку, если пользователя не существует" {
        val ex = shouldThrow<UserNotFoundException> {
            userService.getUserByUsername("notExistingUsername")
        }

        ex.message shouldBe "Пользователь с username notExistingUsername не найден"
    }

    //----------------updateUser-----------------

    "updateUser обновляет username, email и password" {
        val userDetails = mockk<org.springframework.security.core.userdetails.UserDetails>()
        every { userDetails.username } returns "email@email.com"

        val dto = UserUpdateRequest(
            username = "newUsername",
            email = "new@email.com",
            password = "newPassword"
        )

        every { repository.findByEmail("email@email.com") } returns Optional.of(user)
        every { repository.findByUsername("newUsername") } returns Optional.empty()
        every { repository.findByEmail("new@email.com") } returns Optional.empty()
        every { passwordEncoder.encode("newPassword") } returns "encodedPassword"

        val result = userService.updateUser(userDetails, dto)

        result shouldBe userResponseLong

        user.username shouldBe "newUsername"
        user.email shouldBe "new@email.com"
        user.password shouldBe "encodedPassword"
    }


    "updateUser выбрасывает ошибку, если username уже существует" {
        val userDetails = mockk<org.springframework.security.core.userdetails.UserDetails>()
        every { userDetails.username } returns "email@email.com"

        val dto = UserUpdateRequest(
            username = "existingUsername",
            email = null,
            password = null
        )

        every { repository.findByEmail("email@email.com") } returns Optional.of(user)
        every { repository.findByUsername("existingUsername") } returns Optional.of(user)

        val ex = shouldThrow<UserAlreadyExistsException> {
            userService.updateUser(userDetails, dto)
        }

        ex.message shouldBe "Пользователь с username existingUsername уже существует"
    }

    "updateUser выбрасывает ошибку, если email уже существует" {
        val userDetails = mockk<org.springframework.security.core.userdetails.UserDetails>()
        every { userDetails.username } returns "email@email.com"

        val dto = UserUpdateRequest(
            username = null,
            email = "existing@email.com",
            password = null
        )

        every { repository.findByEmail("email@email.com") } returns Optional.of(user)
        every { repository.findByEmail("existing@email.com") } returns Optional.of(user)

        val ex = shouldThrow<UserAlreadyExistsException> {
            userService.updateUser(userDetails, dto)
        }

        ex.message shouldBe "Пользователь с email existing@email.com уже существует"
    }

    //----------------deleteUser-----------------

    "deleteUser удаляет пользователя" {
        val userDetails = mockk<org.springframework.security.core.userdetails.UserDetails>()
        every { userDetails.username } returns "email@email.com"

        every { repository.findByEmail("email@email.com") } returns Optional.of(user)
        every { repository.delete(user) } returns Unit

        userService.deleteUser(userDetails)

        verify(exactly = 1) { repository.delete(user) }
    }
})